### ISSUE TITLE
simplemodbus pip package discontinued?

### ISSUE THREAD

**Issue by MrRusch** [larsvinc/swegon-to-mqtt#1]
I will be attempting to integrate my Swegon Casa R3 Smart and Openhab2 setup.
Like you, I'm doing this through a RPI Zero W, through a [USB2RS485 dongle](https://hitechchain.se/arduinokompatibel/industrial-usb-to-rs485-converter)

When trying your setup script, I am met by

> Collecting simplemodbus
> Could not install packages due to an EnvironmentError: 404 Client Error: Not Found for url: https://pypi.org/simple/simplemodbus/

..and the folder does indeed seem to be missing from https://pypi.org/ so I guess it has been discontinued?

What are the chances you will update your project with a more recent package?

EDIT:
OK so looking at the imports you are using minimalmodbus package.. So I guess it's just the setup script that is out of date then, easy fix.

Best regards

**Comment by larsvinc** [comment]
Hi!

Thanks for pointing out. Yes, it is minimalmodbus. It might be a typo from me actually.

What modbus adapter are you using? I think the modbus API is very different between versions. Ideally, the default settings are moved to a settings file, and the driver would detect both the modbus API version and the swegon aparatus and use the appropriate commands?

**Comment by larsvinc** [comment]
Fixed the setup-file.

**Comment by MrRusch** [comment]
> Hi!
> 
> Thanks for pointing out. Yes, it is minimalmodbus. It might be a typo from me actually.
> 
> What modbus adapter are you using? I think the modbus API is very different between versions. Ideally, the default settings are moved to a settings file, and the driver would detect both the modbus API version and the swegon aparatus and use the appropriate commands?

I have ordered Swegon's [SEC connection module](https://www.swegon.com/products/air-handling/casa/accessories/control-and-regulating-equipment/casa-sec/). The documentation there refers to device version 3.0 for the registers listed.
What you describe sounds great. Not sure I will be able to produce it though 😉

**Comment by larsvinc** [comment]
I see. Note that this driver works with API version 1.2. I am in contact with swegon to see if I can get an updated set of registers to make it work for 3.0.

**Comment by MrRusch** [comment]
> I see. Note that this driver works with API version 1.2. I am in contact with swegon to see if I can get an updated set of registers to make it work for 3.0.

I assumed it was the ones listed under documentation for the SEC:
https://www.swegon.com/globalassets/_product-documents/home-ventilation/accessories/_multi/smart_modbus_register.pdf
I also, probably falsely, assumed that the Modbus version was decided by the SEC. But now I'm thinking this should be decided by the module integrated in the fan..? Can't find any information on what Modbus version the fan is using in its documentation:
https://www1.swegon.com/Global/PDFs/Home%20ventilation/Air%20handling%20units/Swegon%20CASA%20R-series/_en/R3_f_EN-t_web.pdf

**Comment by larsvinc** [comment]
There are many API versions (1.2, 2, 3, 3.1) that I have seen, and most of them have massive breaking changes. I am looking to update to version 2 now.

Let me know what you find. And how the code must be modified to handle the differences.

**Comment by MrRusch** [comment]
OK so after having waited a couple of weeks for a response per email I called tech support and spoke to someone who seemed very annoyed by my question. He said that the unit implemented version 3.0 of the Modbus API. But he also said that the SEC module I have installed won't work and that I need the SEM module. According to manual however the SEC module is intended for single point integration and SEM is for large networks. So I'm not sure what to think.. But I guess maybe try reading some registers according to 3.0 specification? I've got everything I need hardware wise now so will start tinkering soon.

**Comment by larsvinc** [comment]
@MrRusch sounds good.

I have had some issues with my system, and support has now given me a new gateway. Then I will implement version 2.0 of the Modbus API. It is however quite some effort since version 2.0 is a completely different API. So I expect to finalize this in March.

**Comment by MrRusch** [comment]
@larsvinc Awesome. Hopefully I can make some contributions along the way.

I have just done some successful communication tests. 

Turns out, the ventilation company had installed my SEC in the wrong port (SET 1).. Was not getting any response from the unit during last nights tests. Checked the manual, checked the wiring, checked the cables.. Ended up opening the unit up and voila - now everything works fine. 

Time to dig into your code.

**Comment by larsvinc** [comment]
Great!

A few notes:
- the code is compatible with version 1.2
- In particular, there is both a clock that I have to monitor for malfunction, and more fan-modes than in the more modern systems

**Comment by paulianttila** [comment]
@larsvinc I have also reset issue with my Swegon Casa W100. When modbus gw is installed (v1.2), W100 reset randomly. Did you get never modbus gw version? Did it solve the reset issue?

**Comment by larsvinc** [comment]
@paulianttila I never got this issue resolved. Is the issue when you connect the Gateway (with our without a modbus connection)?

**Comment by paulianttila** [comment]
Reset happens when modbus gw is connected to Swegon Casa with or without modbus client  connection/communication.

**Comment by paulianttila** [comment]
This is now fully offtopic related to this issue, but as I have now found fate mate I continue the discussion :) 

My understanding so far have been that I have faulty modbus gw, so I have not used it. Therefore I have implemented app which receive data directly from ventilation unit bus and parse it (reverse engineered the protocol a bit). This app doesn't ever send anything to the ventilation system, but just receive live data and then I do all controls direclty via IO. See [org.openhab.binding.swegonventilation](https://github.com/openhab/openhab1-addons/tree/main/bundles/binding/org.openhab.binding.swegonventilation) I don't use openHAB 1 binding anymore as I have also implemented mqtt gw solution, where data is published to the mqtt. This app also have a analyzer functioanlity to see the communication in the bus.

As energy prices are now going up, I like to control the unit better than what is possible via IO as it rather limited. Now as I found that someone else have also reset issue with modbus unit, I started the analyse it bit further. So what I did, I installed modbus unit and I can see random resets again. I started the anylyzer to see the communication between ventilation unit, modbus unit and remote control panel. Communication is rather intensive, around 6 messages per second, so there is a lot of log to analyse as reset can happens only couple times per day. Therefore I implemented a monitoring, so when every ventilation unit resets (time is set to 0) I stop the logging to see easily what happened in the bus during reset.

Control panel doesn't never send anything to ventilation unit if user doesn't change something from the display, but modbus gw is continuously communicating to the ventilation unit. It ask more information all the time. The protocol between ventilation unit and peripheral devices is rather simple. Ventilation unit is master and it "broadcast" live data all the time to the bus and then ones a while query if some of the peripheral devices have anything to send. Modbus gw always ask couple of other message types, which ventilation unit will then send to the modbus gw. Modbus gw ask more data ones in every couple of seconds. PDU which modbus gw use to ask more data is static, so it basically just ask 10 or 11 message type (10 and 11 messages are never broadcasted). So what now happend in the reboot situation, modbus gw ask more data, ventilation unit sent message(s) and then reboot. So I would blame the ventilation unit. It probably have software bug which cause the reboot.

**Comment by larsvinc** [comment]
@paulianttila Fantastic to meet a fate mate. I probably burned through a ton of hours debugging this! Ran my rpi on a power bank to avoid earth loops! Disconnected the user panel, changed the user panel, firmware upgrades etc. I am not surprised that it is the gateway. I am tempted to contact Swegon and ask for a new one. At the same time, implementing this over again is a major hazzle. So thinking about it!

Very impressed by your debugging!

On a note, the library that I wrote here has been running for years. I get about 20 resets per day, and I can hear the ventilation rebooting, but overall it works well and gives me the control I want. I doubt that Swegon will fix bugs in such an old, depricated platform, but perhaps it is worth a try...
